contracts/Comptroller/Diamond/Diamond.sol

```diff
@@ -0,0 +1,200 @@
+        // add new facet address if it does not exist
+        if (selectorPosition == 0) {
+            addFacet(_facetAddress);
+        }
+        for (uint256 selectorIndex; selectorIndex < _functionSelectors.length; selectorIndex++) {
+            bytes4 selector = _functionSelectors[selectorIndex];
+            address oldFacetAddress = selectorToFacetAndPosition[selector].facetAddress;
+            require(oldFacetAddress != _facetAddress, "LibDiamondCut: Can't replace function with same function");
+            removeFunction(oldFacetAddress, selector);
+            addFunction(selector, selectorPosition, _facetAddress);
+            selectorPosition++;
+        }
+    }
+
+    /**
+     * @notice Remove function selectors to the facet's address mapping.
+     * @param _facetAddress Address of the facet.
+     * @param _functionSelectors Array of function selectors need to remove in the mapping.
+     */
+    function removeFunctions(address _facetAddress, bytes4[] memory _functionSelectors) internal {
+        require(_functionSelectors.length > 0, "LibDiamondCut: No selectors in facet to cut");
+        // if function does not exist then do nothing and return
+        require(_facetAddress == address(0), "LibDiamondCut: Remove facet address must be address(0)");
+        for (uint256 selectorIndex; selectorIndex < _functionSelectors.length; selectorIndex++) {
+            bytes4 selector = _functionSelectors[selectorIndex];
+            address oldFacetAddress = selectorToFacetAndPosition[selector].facetAddress;
+            removeFunction(oldFacetAddress, selector);
+        }
+    }
+
+    /**
+     * @notice Add new facet to the proxy.
+     * @param _facetAddress Address of the facet.
+     */
+    function addFacet(address _facetAddress) internal {
+        enforceHasContractCode(_facetAddress, "Diamond: New facet has no code");
+        facetFunctionSelectors[_facetAddress].facetAddressPosition = facetAddresses.length;
+        facetAddresses.push(_facetAddress);
+    }
+
+    /**
+     * @notice Add function selector to the facet's address mapping.
+     * @param _selector funciton selector need to be added.
+     * @param _selectorPosition funciton selector position.
+     * @param _facetAddress Address of the facet.
+     */
+    function addFunction(bytes4 _selector, uint96 _selectorPosition, address _facetAddress) internal {
+        selectorToFacetAndPosition[_selector].functionSelectorPosition = _selectorPosition;
+        facetFunctionSelectors[_facetAddress].functionSelectors.push(_selector);
+        selectorToFacetAndPosition[_selector].facetAddress = _facetAddress;
+    }
+
+    /**
+     * @notice Remove function selector to the facet's address mapping.
+     * @param _facetAddress Address of the facet.
+     * @param _selector function selectors need to remove in the mapping.
+     */
+    function removeFunction(address _facetAddress, bytes4 _selector) internal {
```

Comment:
Are there unit tests for the functions in this contract? The logic looks quite complex, IMO it would be useful to test the behavior for arbitrary selectors and corner cases like removing the last selector in the array, etc.

Reply:
Added, 582ae8912c74f4f18f01aa1a85d601f1c88488af